proposals/0029-cooperative-vector.md

```diff
@@ -102,383 +111,336 @@ Introduce new HLSL intrinsics to accelarate matrix-vector operations. In this sp
-(#minimum-support-set).  Note that *Transposed* is always queried.
+> * TODO: are packed types allowed for bias vectors?
 
-Non-"Packed" type interpretations are used to request arithmetic conversions. Input type must be a 32-bit or 16-bit
-scalar integer or a 32-bit or 16-bit float. Integer to integer conversion saturates, float to float conversion is
-implementation dependent and preserves the value as accurately as possible. Float to integer conversion is RTNE and
-saturating. Integer to float conversion is RTNE.
+The bias matrix is loaded from the raw-buffer, **bias vector resource**,
+starting at **bias vector offset**. The **bias vector interpretation** argument
+specifies the element type of the bias vector (see [Type Interpretations]).
 
-/// XXX TODO: These rules make sense for NN applications but diverge from HLSL conversion rules [here]
-    (https://microsoft.github.io/hlsl-specs/specs/hlsl.html#Conv).
+The base address of **bias vector resource** and **bias vector offset** must be
+64 byte aligned.
 
-"Packed" type conversions are bitcasts to a smaller type. The declared input type must be 32-bit unsigned integer. 
 
-/// XXX TODO: Error handling for illegal conversions. 
+### Vector Outer Product
 
-Examples:
+#### Syntax
 
-Packed Case:
+``` llvm
+declare void @dx.op.vecouterproductacc.v[M][TY].v[N][TY](
+    immarg i32,       ; opcode 
+    <[M] x [TY]>,     ; input vector 1
+    <[N] x [TY]>,     ; input vector 2
+    %dx.types.Handle, ; matrix resource
+    i32,              ; matrix offset 
+    i32,              ; matrix stride 
+    immarg i32,       ; matrix interpretation 
+    immarg i32)       ; matrix layout 
 ```
-// Declare an input vector
-vector<uint, 8> ipVector;
-
-// Set interpretation value to DXILCoopVectorTypeInterpretation::SignedInt8x4Packed
-// Each uint element (32-bit) in the input vector, ipVector, will be interpreted as 4 int8 values in the VectorMatrixMul intrinsic. 
-// Note that InputTy = uint and InputComponents = 8 (8 x 4 = 32 sint8 values )
-VecMatOpDescriptor<32 /*M*/, 
-                   32 /*K*/, 
-                   DXILTypeInterpretation::SignedInt8x4Packed /*InputInterpretation*/, 
-                   DXILTypeInterpretation::SignedInt8 /*MatrixInterpretation*/,
-                   DXILTypeInterpretation::Unsupported /*BiasInterpretation*/, 
-                   DXILMatrixLayout::InferencingOptimal /*Layout*/,
-                   false /*Transpose*/> desc;
-
-vector<int, 32> resultVector; //Note that the ResultComponents equals M(32)
-// Matrix is a ByteAddressBuffer
-resultVector = VectorMatrixMul(ipVector, Matrix, 0/*MatrixOffset*/, 0/*MatrixStride*/);
 
-```
+#### Overview
 
-Non-Packed Case:
-```
-// Declare an input vector
-vector<float, 32> ipVector;
+Computes the outer product between column vectors and an **M**x**N** matrix* is accumulated atomically (with device scope) in memory. 
```

Comment:
extra * after matrix, replace "atomically" with "component-wise" atomically. The operation on the whole vector isn't atomic, just the individual components.

Reply:
@anupamachandra - like this?